external/langchaingo/textsplitter/token_splitter.go

```diff
@@ -0,0 +1,82 @@
+package textsplitter
+
+import (
+	"fmt"
+
+	"github.com/pkoukk/tiktoken-go"
```

Comment:
👀 In `pipeline-backend-cloud` we use https://github.com/hupe1980/go-tiktoken, please consider using a single package across our codebase.

Since this is going to be used by an operator, I'd avoid making external calls so, if we _need to_ keep this package, I'd use [this loader](https://github.com/pkoukk/tiktoken-go-loader).

Reply:
@jvallesm 
> Why not importing the package? If not, what are the main differences? 

There is no difference. I chose this way because of this discussion.
- [discussion](https://linear.app/instill-ai/issue/INS-4499/[component][text-operator]-add-task-chunk-text-for-more-advanced#comment-1da16232)
In short, the main reason is that importing all package needs to upgrade Go version.
And, I am not quite sure the risk about updating Go version.
So, I want to keep updating Go version in another branch.

> The project is actively maintained so by duplicating its code we're losing any update they make (e.g. a vulnerability fix).

I agree! I am thinking about we transfer to import the package directly after we upgrade Go version.
However, I am not quite sure about this topic. 
Do you have experience on this part?
- [It has its own data model and interfaces for connecting LLMs, database I/O, etc., which is a bit too heavy for a single Instill Component. We probably want to implement the code ourselves instead of importing it as a whole Go package.](https://linear.app/instill-ai/issue/INS-4499/[component][text-operator]-add-task-chunk-text-for-more-advanced#comment-fd95adca)

Reply:
I checked Task_Split_By_Token, and it used `"github.com/pkoukk/tiktoken-go"`
- [code](https://github.com/instill-ai/component/blob/main/operator/text/v0/split.go#L4)

Is it different between `pipeline-backend-cloud` & `pipeline-backend` now?
Should we modify `pipeline-backend-cloud`?

I am not quite sure about whether [it is a good source](https://github.com/hupe1980/go-tiktoken).
It seems the star / fork count is low. 
Is it a standard to judge?

Reply:
- About the Go version, I'd import an earlier version of the package that is compatible with our Go version. If this is the reason to copy instead of import the code, a much more sensible approach is updating Go, then importing the latest version of the dependency, even if in the beginning we might potentially lose some functionality like a given token encoding. The version difference is small so this shouldn't be a big migration task.
- About the size of `langchaingo`, the rule of thumb I'd follow would be that it's a good argument as long as what you're trying to copy fits in your package (a function, a struct type). Creating 2 `external` packages and importing them adds confusion to the structure and maybe it's a smell that indicates that it's worth importing the source (and more honest). At this point I don't think binary size is a massive concern and I think there are [plenty of ways to improve it](https://instill-ai.slack.com/archives/C01496MQAQJ/p1714469602131749) anyways. In other SDKs like S3 or Stripe we're using a tiny subset of their features and we're not considering copying their source code.

Reply:
> Is it different between pipeline-backend-cloud & pipeline-backend now? Should we modify pipeline-backend-cloud?

`pipeline-backend` doesn't use it because this is needed for the Credit feature, implemented only on Cloud.

> I am not quite sure about whether [it is a good source](https://github.com/hupe1980/go-tiktoken).

I picked the other package (without being aware of `tiktoken-go` being used in `component`) because the encoding is embedded by default and because gpt-4o was quickly added after it was announced, so it fit our needs. The star count is definitely an indicator of the quality and health of the project but we shouldn't be guided solely by it and in this case the choice made sense. I'm not against modifying `pipeline-backend-cloud` instead of this package.

Reply:
@jvallesm 
> even if in the beginning we might potentially lose some functionality like a given token encoding. The version difference is small so this shouldn't be a big migration task.

Just share
In my previous experience / industry / company, losing functionality is not acceptable in terms of client sides. So, we barely upgrade the language version to keep the stability.

About Go version, I checked the [release note](https://tip.golang.org/doc/go1.22).
It seems like what you said, it is not a big migration.
However, I still want to separate branches between upgrading Go version & this feature branch.
I think it will be clearer than combining them together.

How about we mark it as TODO and import the package after we upgrade the Go version?
What do you think about it?

>  I'm not against modifying pipeline-backend-cloud instead of this package.

I am confused about this. So, I want to confirm it again.
If we are going to import all langchaingo package, should we modify `pipeline-backend-cloud` to avoid two token package rather than `pipeline-backend`?

Reply:
@chuang8511 
I'll upgrade the Golang version today in another PR.

If my understanding is correct, after we upgrade the Golang version, we don't need to modify any Golang code from LangChainGo, right? If so, let's just import it.

[It has its own data model and interfaces for connecting LLMs, database I/O, etc., which is a bit too heavy for a single Instill Component. We probably want to implement the code ourselves instead of importing it as a whole Go package.](https://linear.app/instill-ai/issue/INS-4499/%5Bcomponent%5D%5Btext-operator%5D-add-task-chunk-text-for-more-advanced#comment-fd95adca)
This only happens when you need to make modifications. If we don't update their code, we can just import it. The size of the package is not a problem, I remember that when using `go build`, the compiler will only compile the code you used.

Reply:
I just checked https://github.com/pkoukk/tiktoken-go, and they have added GPT-4 support. I guess we can safely switch to it.